lib/propolis/src/accessors.rs

```diff
@@ -71,93 +101,81 @@ impl<T> Tree<T> {
             .get_mut(&parent)
             .expect("parent node must exist")
             .children
-            .insert(child);
+            .insert(child_key);
+
+        child_node
     }
 
-    fn adopt(
-        &mut self,
-        leaf_ent: &NodeEntry<T>,
-        adopt_tree: &mut Tree<T>,
-        mut adopt_root: MutexGuard<NodeEntry<T>>,
-    ) {
+    /// Adopt the root node and all its descendants into our tree, under the
+    /// node specified by `parent_key`
+    fn adopt(&mut self, parent_key: NodeKey, adopt_tree: &mut Tree<T>) {
         debug_assert!(
             self.nodes
-                .get(&leaf_ent.id)
-                .and_then(|n| Weak::upgrade(&n.node_ref))
+                .get(&parent_key)
+                .and_then(|node| Weak::upgrade(&node.node_ref))
                 .is_some(),
             "leaf target for re-parenting missing"
         );
-        assert_eq!(adopt_root.id, ID_ROOT);
-        let tree_ref = &leaf_ent.tree;
+
+        let child_key = adopt_tree.root_key;
+        let tree_ref = self.self_weak.upgrade().unwrap();
 
         let mut queue = VecDeque::new();
-        queue.push_back((ID_ROOT, leaf_ent.id));
-        while let Some((child_id, local_parent_id)) = queue.pop_front() {
-            if let Some(mut tnode) = adopt_tree.nodes.remove(&child_id) {
+        queue.push_back(child_key);
+        while let Some(adopt_key) = queue.pop_front() {
+            if let Some(mut tnode) = adopt_tree.nodes.remove(&adopt_key) {
                 let node = match Weak::upgrade(&tnode.node_ref) {
                     Some(nr) => nr,
                     None => {
                         continue;
                     }
                 };
 
-                let new_child_id = self.next_id();
-
-                if child_id != ID_ROOT {
-                    let mut ent = node.0.lock().unwrap();
-                    // Place the node in our tree
-                    debug_assert_eq!(ent.id, child_id);
-                    ent.id = new_child_id;
-                    ent.tree = Arc::clone(tree_ref);
-                    ent.resource = self.resource_root.clone();
-                    drop(ent);
-                } else {
-                    // Processing for the child root node is special, as we
-                    // already hold the lock on it.
-                    debug_assert_eq!(adopt_root.id, child_id);
-                    adopt_root.id = new_child_id;
-                    adopt_root.tree = Arc::clone(tree_ref);
-                    adopt_root.resource = self.resource_root.clone();
-                }
+                let mut ent = node.0.lock().unwrap();
+                ent.tree = Arc::clone(&tree_ref);
+                ent.resource = self.resource_root.clone();
+                drop(ent);
```

Comment:
silly style nit, feel free to disagree: in general, I prefer to use scopes with curly-braces for stuff like this, rather than explicit `drop`s, it makes the portion of code for which the lock is held a little more visually obvious to readers.

also, i notice that we do it that way later in the file, so the consistency might be nice?

Reply:
This should be cleaned up, as well as the other instances I found.  (I didn't bother going through the tests, though)